rosetta/transactions/sign_transaction.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package transactions
+
+import (
+	"encoding/hex"
+
+	"github.com/onflow/flow-go-sdk"
+
+	"github.com/optakt/flow-dps/rosetta/failure"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+// SignTransaction adds the given signature to the transaction.
+func (p *Parser) SignTransaction(tx *flow.Transaction, signature object.Signature) (*flow.Transaction, error) {
+
+	// Validate the transaction actors. We expect a single authorizer - the sender account.
+	if len(tx.Authorizers) != authorizersRequired {
+		return nil, failure.InvalidAuthorizers{
+			Have:        uint(len(tx.Authorizers)),
+			Want:        authorizersRequired,
+			Description: failure.NewDescription("invalid number of authorizers"),
+		}
+	}
+
+	sender := tx.Authorizers[0]
+
+	// Verify that the sender is the payer, since it is the payer that needs to sign the envelope.
+	if tx.Payer != sender {
+		return nil, failure.InvalidPayer{
+			Have:        tx.Payer,
+			Want:        sender,
+			Description: failure.NewDescription("invalid transaction payer"),
+		}
+	}
+
+	// Verify that we do not already have signatures.
+	if len(tx.EnvelopeSignatures) > 0 {
+		return nil, failure.InvalidSignature{
+			Description: failure.NewDescription("unexpected envelope signatures found",
+				failure.WithInt("signatures", len(tx.EnvelopeSignatures))),
+		}
+	}
+
+	// Verify that the signature belongs to the sender.
+	signer := flow.HexToAddress(signature.SigningPayload.AccountID.Address)
+	if signer != sender {
+		return nil, failure.InvalidSignature{
+			Description: failure.NewDescription("invalid signer account",
+				failure.WithString("have_signer", signer.String()),
+				failure.WithString("want_signer", sender.String())),
+		}
+	}
+
+	bytes, err := hex.DecodeString(signature.HexBytes)
+	if err != nil {
+		return nil, failure.InvalidSignature{
+			Description: failure.NewDescription("invalid signature payload",
+				failure.WithErr(err)),
+		}
+	}
+
+	// Copy the transaction and add the signature.
+	copyTx := *tx
```

Comment:
This is _not_ how you copy a Go struct with pointer semantics. All this will do is copy all the members that are value types. However, for slices and maps, you will copy references to the same slices and maps. For pointers, you will copy a pointer that points to the same underlying struct.

Reply:
If there is no `Copy` or `Clone` on the Flow transactions, then an easy way to do it is to marshal/unmarshal. Or you can use a library, such as https://github.com/barkimedes/go-deepcopy.

Reply:
I had it in my head that it's not really okay, but I thought because of the short lifespan of the original transaction, it might be okay.

Still, it's an expensive gamble and not really okay.

I think the new way is pretty nice, it's in the same vein as the marshal/unmarshal you mentioned.